keras/backend/theano_backend.py

```diff
@@ -414,25 +416,34 @@ def rnn(step_function, inputs, initial_states,
             the step function, of shape (samples, ...).
     '''
     inputs = inputs.dimshuffle((1, 0, 2))
+    if mask is None:
+        mask = expand_dims(ones_like(T.sum(inputs, axis=-1)))
+    else:
+        mask = mask.dimshuffle((1, 0, 2))
 
-    def _step(input, *states):
+    def _step(input, mask, output_tm1, *states):
         output, new_states = step_function(input, states)
-        if masking:
-            # if all-zero input timestep, return
-            # all-zero output and unchanged states
-            switch = T.any(input, axis=-1, keepdims=True)
-            output = T.switch(switch, output, 0. * output)
-            return_states = []
-            for state, new_state in zip(states, new_states):
-                return_states.append(T.switch(switch, new_state, state))
-            return [output] + return_states
-        else:
-            return [output] + new_states
+        output = T.switch(mask, output, output_tm1) # output previous output if masked.
+        return_states = []
+        for state, new_state in zip(states, new_states):
+            return_states.append(T.switch(mask, new_state, state))
+        return [output] + return_states
+
+    # build an all-zero tensor of shape (samples, output_dim)
+    # (aside, it's kind of crazy that we do this in this way, if the compiler
+    # isn't smart about the outer product below, it could be relatively expensive.
+    initial_output = zeros_like(inputs)  # (timesteps, samples, input_dim)
+    initial_output = sum(initial_output, axis=0)  # (samples, input_dim)
+    initial_output = sum(initial_output, axis=1)  # (samples,)
+    # I'm putting the extra zeros_like here in the *hopes* that it helps the optimizer
+    # realize that it never needs to compute the actual outer product but that it's just for
+    # shape inference.
+    initial_output = zeros_like(T.outer(initial_output, zeros((output_dim,)))) # (samples, output_dim)
 
     results, _ = theano.scan(
         _step,
-        sequences=inputs,
-        outputs_info=[None] + initial_states,
+        sequences=[inputs, mask],
+        outputs_info=[initial_output] + initial_states,
```

Comment:
I agree there's no conceptual reason why this shouldn't be possible. Do you have a more specific idea on how to hack around the Theano API issue? Even to feed in dummy data to the step function I need to know the correct dimensions of the dummy data. If I try to `eval` the shape of the inputs matrix I get a disconnected input error because it's not yet hooked up at compile time.